File: 0018-4sum/0018-4sum.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
          sort(nums.begin(),nums.end());
        vector<vector<int>>ans; int st=0, end=0;
        long long int elem =0 ;
         if(nums.size() <4) return ans;
        set<vector<int>>set;
        for(int i=0;i<nums.size()-2;i++){
            // if(i>0 && nums[i] == nums[i-1]){
            //     continue;
            // }
            for(int j=i+1;j<nums.size()-1;j++){
            //      if(j>0 && nums[j] == nums[j-1]){
            //     continue;
            // }
            st = j+1;end = nums.size()-1;
                while(st < end){
                   elem = (nums[i] + nums[j]) ;
                   if(target>(elem+nums[st] + nums[end])){
                    st++;
                   }
                   else if(target<(elem + nums[st] + nums[end])) end--;
                   else{
                    //ans.push_back({nums[i],nums[j],nums[st],nums[end]});
                    set.insert({nums[i],nums[j],nums[st],nums[end]});
                    st++; end--;
                    while(nums[st] == nums[st-1] && st < end){
                      st++;
                    }
                   }
                }
            }
        }
       for(auto x : set){
         ans.push_back(x);
       }
        return ans;
    }
};

File: 0018-4sum/0018-4sum.java
class Solution {
    public List<List<Integer>> fourSum(int[] nums, int target) {
         Arrays.sort(nums);
         List<List<Integer>> ans = new ArrayList<>();
         Set<List<Integer>>s = new HashSet<>();

         for(int j=0;j<nums.length;j++){
             for(int i=j+1;i<nums.length;i++){
           int st = i + 1, end = nums.length -1;
           if(st>=nums.length || end>= nums.length) continue;
          while(st < end){
            long sum = 0; sum += nums[i];
            sum += nums[j];
            sum+= nums[st]; 
            sum+= nums[end];
            if(sum == (long)target){
                List<Integer> ls = new ArrayList<>();
                ls.add(nums[i]);
                 ls.add(nums[j]);
                ls.add(nums[st]);
                ls.add(nums[end]);
                s.add(ls);
                st++;
                end--;
            }
            else if(sum > (long)target){
                end--;
            }
            else st++;
          }
        }

         }
        

        for(List<Integer> ls: s){
           ans.add(ls);
        }
     return ans;
    }
}